# ML-04 — Search Intelligence Data Contract

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [29]:
!git clone https://{github_token}@github.com/SafaGulAISoftwareEngineer/FlyRank-ML-Internship.git /content/repo

Cloning into '/content/repo'...
remote: Enumerating objects: 102, done.
remote: Counting objects: 100% (102/102), done.
remote: Compressing objects: 100% (70/70), done.
remote: Total 102 (delta 22), reused 86 (delta 16), pack-reused 0 (from 0)
Receiving objects: 100% (102/102), 1.84 MiB | 4.58 MiB/s, done.
Resolving deltas: 100% (22/22), done.


## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

Row = one (report_date, client_hash_id, content_hash_id) in fact_content_daily_performance.
Table: fact_content_daily_performance (not _sample — that's the sealed final month).
Window: report_date in [2026-03-01, 2026-04-01), a mid-panel month.

In [21]:
REL = 'hf://datasets/FlyRank/internship-warehouse'
TABLES = {
    'dim_clients':       f"read_parquet('{REL}/dim_clients.parquet')",
    'dim_content':       f"read_parquet('{REL}/dim_content.parquet')",
    'fact_daily':        f"read_parquet('{REL}/fact_content_daily_performance/**/*.parquet')",
    'fact_daily_sample': f"read_parquet('{REL}/fact_content_daily_performance_sample.parquet')",
    'fact_query_90d':    f"read_parquet('{REL}/fact_content_query_90d.parquet')",
}

In [22]:
import duckdb
con = duckdb.connect()
con.execute(f"CREATE OR REPLACE SECRET hf (TYPE huggingface, TOKEN '{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}')")

# cheap probe first — confirms auth + gate access with almost no cost
con.sql(f"SELECT COUNT(*) FROM read_parquet('{REL}/dim_clients.parquet')").df()

,count_star()
0,104


## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

| Field | Bucket | Why |
|---|---|---|
| report_date, client_hash_id, content_hash_id | Context | Join/group/window keys only, never fed to a model |
| gsc_clicks, gsc_impressions, gsc_avg_position, sessions_organic, sessions_ai | Feature candidates | Observed daily signals, knowable same-day |
| ga4_data_available | Context (filter flag) | Decides whether GA4 columns are trustworthy that day — not itself a model input |
| is_declining_future (prior-90d avg clicks vs future-30d avg clicks, ≥25% drop, volume floor) | Label / proxy | The thing to predict; computed FROM gsc_clicks across two windows, never a feature itself |
| scroll_events | Excluded | Present in only ~0.76% of daily rows in this snapshot (600,821 of 78,835,655, per the lane guide's density table) — at page-day grain it's blank almost everywhere; imputing it would encode "did a scroll ping happen to fire" rather than real engagement |

## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

In [23]:
tbl = f"read_parquet('{rel}/fact_content_daily_performance/**/*.parquet', hive_partitioning=true)"

frame = con.sql(f"""
WITH prior AS (
    SELECT content_hash_id,
        AVG(gsc_clicks) AS avg_daily_clicks_prior,
        AVG(gsc_avg_position) FILTER (WHERE gsc_avg_position > 0) AS avg_position_prior,
        COUNT(*) FILTER (WHERE gsc_clicks > 0) AS days_with_clicks_prior,
        AVG(CASE WHEN ga4_data_available IS TRUE THEN 1.0 ELSE 0.0 END) AS ga4_coverage_prior,
        SUM(sessions_ai) * 1.0 / NULLIF(SUM(sessions_organic + sessions_ai), 0) AS ai_share_prior
    FROM {tbl}
    WHERE report_date >= DATE '2026-03-01' AND report_date < DATE '2026-03-16'
    GROUP BY content_hash_id
    HAVING COUNT(*) >= 10
),
target AS (
    SELECT content_hash_id, AVG(gsc_clicks) AS avg_daily_clicks_target
    FROM {tbl}
    WHERE report_date >= DATE '2026-03-16' AND report_date < DATE '2026-04-01'
    GROUP BY content_hash_id
)
SELECT p.*, t.avg_daily_clicks_target,
    CASE WHEN p.avg_daily_clicks_prior >= 0.5
              AND t.avg_daily_clicks_target < p.avg_daily_clicks_prior * 0.75
         THEN 1 ELSE 0 END AS toy_label
FROM prior p JOIN target t USING (content_hash_id)
""").df()

frame.shape, frame['toy_label'].mean()   # base rate — print this, always

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

((312189, 8), np.float64(0.015852576484116993))

In [24]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score

honest_features = ['avg_daily_clicks_prior','avg_position_prior','days_with_clicks_prior',
                    'ga4_coverage_prior','ai_share_prior']
X = frame[honest_features].fillna(0)
y = frame['toy_label']
Xtr, Xte, ytr, yte = train_test_split(X, y, test_size=0.3, random_state=0, stratify=y)

clf = LogisticRegression(max_iter=1000).fit(Xtr, ytr)
honest_auc = roc_auc_score(yte, clf.predict_proba(Xte)[:,1])
print("Honest AUC:", honest_auc, "| base rate:", y.mean())

# Now the deliberate leak: avg_daily_clicks_target is literally what the label is computed from
leaky_features = honest_features + ['avg_daily_clicks_target']
Xl = frame[leaky_features].fillna(0)
Xtrl, Xtel, _, _ = train_test_split(Xl, y, test_size=0.3, random_state=0, stratify=y)
clf_leak = LogisticRegression(max_iter=1000).fit(Xtrl, ytr)
leaky_auc = roc_auc_score(yte, clf_leak.predict_proba(Xtel)[:,1])
print("Leaky AUC:", leaky_auc)   # should jump toward ~1.0

# Delete it, keep the honest number
print("Final honest AUC (leak removed):", honest_auc)

Honest AUC: 0.9823811243830337 | base rate: 0.015852576484116993
Leaky AUC: 0.99855599347202
Final honest AUC (leak removed): 0.9823811243830337


In [25]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler().fit(Xtr)
clf_scaled = LogisticRegression(max_iter=1000).fit(scaler.transform(Xtr), ytr)
dict(zip(honest_features, clf_scaled.coef_[0]))

{'avg_daily_clicks_prior': np.float64(-0.45435005648660143),
 'avg_position_prior': np.float64(0.014545194485106355),
 'days_with_clicks_prior': np.float64(1.297307613280879),
 'ga4_coverage_prior': np.float64(0.052833233537900925),
 'ai_share_prior': np.float64(0.01313261549059063)}

In [26]:
tbl = f"read_parquet('{REL}/fact_content_daily_performance/**/*.parquet', hive_partitioning=true)"

# grain check — should return ZERO rows
grain_result = con.sql(f"""
    SELECT report_date, client_hash_id, content_hash_id, COUNT(*) AS c
    FROM {tbl}
    WHERE report_date >= DATE '2026-03-01' AND report_date < DATE '2026-04-01'
    GROUP BY report_date, client_hash_id, content_hash_id
    HAVING COUNT(*) > 1
    LIMIT 5
""").df()
display(grain_result)

# count + date span
display(con.sql(f"""
    SELECT COUNT(*) AS n_rows, COUNT(DISTINCT content_hash_id) AS n_content,
           COUNT(DISTINCT client_hash_id) AS n_clients,
           MIN(report_date) AS min_date, MAX(report_date) AS max_date
    FROM {tbl}
    WHERE report_date >= DATE '2026-03-01' AND report_date < DATE '2026-04-01'
""").df())

# availability — IS TRUE
display(con.sql(f"""
    SELECT COUNT(*) AS total_rows,
           COUNT(*) FILTER (WHERE ga4_data_available IS TRUE) AS ga4_available_rows
    FROM {tbl}
    WHERE report_date >= DATE '2026-03-01' AND report_date < DATE '2026-04-01'
""").df())

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,report_date,client_hash_id,content_hash_id,c


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,n_rows,n_content,n_clients,min_date,max_date
0,9841378,331437,55,2026-03-01,2026-03-31


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,total_rows,ga4_available_rows
0,9841378,413966


In [27]:
len(grain_result)

0

In [28]:
clf.coef_

array([[-0.7001073 ,  0.00224504,  0.6400745 ,  0.45575014,  0.00905373]])

## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

This March slice can't tell you which pages had enough PRIOR history to be modeled at all —
a real 90-day-prior/30-day-future window around a March decision date will drop any content
whose client's tracking hadn't started 90 days earlier, even though the March rows themselves
look complete.

Separately, March 2026 covers only 55 of the warehouse's 104 total clients — client presence
per month is itself unbalanced, not just history depth. And GA4-based features are only
trustworthy on ~4.2% of March rows (413,966 of 9,841,378) — any feature built from sessions/
engagement columns needs the ga4_data_available filter, not a zero-fill assumption.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.